### PR TITLE
Fix bugs found by mutating source in SwiftParserTest using alternative token choices

### PR DESCRIPTION
**Pull request by ahoppen**
This is a collection of fixes for bugs that I found using the technique in https://github.com/apple/swift-syntax/pull/1340.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftParserTest/ExpressionTests.swift [hunk 1]
```diff
@@ -1652,4 +1652,18 @@ final class StatementExpressionTests: XCTestCase {
...
+  func testUnterminatedInterpolationAtEndOfMultilineStringLiteral() {
+    AssertParse(
+      #"""
+      """\({(1️⃣})
+      2️⃣"""3️⃣
+      """#,
+      diagnostics: [
+        DiagnosticSpec(locationMarker: "1️⃣", message: "expected value and ')' to end tuple"),
+        DiagnosticSpec(locationMarker: "2️⃣", message: #"unexpected code '"""' in string literal"#),
+        DiagnosticSpec(locationMarker: "3️⃣", message: #"expected '"""' to end string literal"#),
```
Comment (bnbarham) [inline Tests/SwiftParserTest/ExpressionTests.swift:1665]: Feels like the `"""` should be a strong marker that an interpolation is unterminated. But I'm all for this vs an infinite loop 😅.

File: Sources/SwiftParser/Declarations.swift [hunk 2]
```diff
@@ -1612,11 +1613,12 @@ extension Parser {
       modifier = nil
     }
 
-    let introducer = self.eat(handle)
```
Comment (bnbarham) [inline Sources/SwiftParser/Declarations.swift:1615]: Whoops!

**Review by bnbarham — APPROVED** [review]

**Comment by bnbarham** [comment]
Most of these seem worth cherry-picking to 5.8 (the infinite loop ones at the very least).

**Pull request merged**